core/consensus/babe/src/lib.rs

```diff
@@ -629,6 +635,34 @@ impl<B: Block, C> Verifier<B> for BabeVerifier<C> where
 					auxiliary: Vec::new(),
 					fork_choice: ForkChoiceStrategy::LongestChain,
 				};
+				const NANOS_PER_SEC: u32 = 1_000_000_000;
+				let mut timestamps = self.timestamps.lock();
+				// Remainder of this block is a critical section.
+				let num_timestamps = timestamps.1.len();
+				if num_timestamps as u64 >= self.config.0.median_required_blocks {
+					let mut new_list: Vec<_> = timestamps.1.iter().map(|&(t, sl)| {
+							let offset: u128 = u128::from(self.config.get())
+								.checked_mul(1_000_000u128) // self.config.get() returns *milliseconds*
+								.and_then(|x| x.checked_mul(u128::from(slot_num).saturating_sub(u128::from(sl))))
+								.expect("we cannot have timespans long enough for this to overflow; qed");
+							let nanos = (offset % u128::from(NANOS_PER_SEC)) as u32;
+							let secs = (offset / u128::from(NANOS_PER_SEC)) as u64;
+							t + Duration::new(secs, nanos)
+						}).collect();
+					// FIXME use a selection algorithm instead of a full sorting algorithm.
+					new_list.sort_unstable();
+					let &median = new_list
+						.get(num_timestamps / 2)
+						.expect("we have at least one timestamp, so this is a valid index; qed");
+					drop(new_list);
+					std::mem::replace(&mut timestamps.1, Default::default());
```

Comment:
```suggestion
					timestamps.1.clear()
```

Is probably much more performant, as we can reuse the vector the next time.